cmd/flux/suspend.go

```diff
@@ -34,22 +34,25 @@ var suspendCmd = &cobra.Command{
 }
 
 type SuspendFlags struct {
-	all bool
+	all    bool
+	reason string
 }
 
 var suspendArgs SuspendFlags
 
 func init() {
 	suspendCmd.PersistentFlags().BoolVarP(&suspendArgs.all, "all", "", false,
 		"suspend all resources in that namespace")
+	suspendCmd.PersistentFlags().StringVarP(&suspendArgs.reason, "reason", "r", "suspended",
```

Comment:
We would add this annotation only of a reason is provided.

Reply:
While I see it beneficial to support the suspend event and optional reason metadata percolate out through flux events, that's outside the scope of this PR.

The default of `--reason suspended` tightly coupling flux cli initiated suspensions to a resource annotation was an oversight of mine.  I'm glad to have such a proper review @darkowlzz @stefanprodan :) and will change this.

Reply:
During the community meeting today, I voiced concerns about this detachment and the possibility that the reason annotation would continue to exist when the `.spec` would move on. For example, when e.g. the `.spec.suspend` boolean is quickly changed in Git without making use of the Flux CLI.

Given this, I am wondering if we could not tackle two issues at once by using `suspend.fluxcd.io/reason` as _another_ way to mark the resource as suspended. This would force a user to remove the annotation to continue the reconciliation of the object (avoiding any potential detachment), while also allowing the object to be suspended without creating a bump in the generation of the object (solving a historical regret around including the suspend field in the spec of the object).

This would raise some questions around how we would e.g. display this in a `kubectl get` column, but does prepare us for a better future in which we could eventually deprecate the `.spec.suspend` in a new major API version.

Reply:
Thinking more about this, it may not even explicitly have to be `/reason`. The mechanics could also be that the presence of the key in the annotation map simply marks the resource as suspended, while the value can be arbitrary. To allow an explicit false, we could e.g. recognize `false | 0 | False` as "disabled" while any other value would mean "enabled".

Reply:
@hiddeco for this feature I originally chased down another fundamental bit a metadata, the `reconcile.fluxcd.io/requestedAt` annotation, as a pattern to follow.  This control point is registered under each resource as a status condition with the flux cli `reconcile` command and each of the controllers all using setter/getter functionality from the `pkg` library to manipulate this annotation (if I'm grokking it).

Considering a future without `spec.suspend` would we consider implementing something akin to this for the suspend metadata?  I think we could have this a parallel code path until such a time that the `.spec.suspend` is deprecated.  The trick would be keeping the spec and metadata in two directional lockstep.

Reply:
> During the community meeting today, I voiced concerns about this detachment and the possibility that the reason annotation would continue to exist when the `.spec` would move on. For example, when e.g. the `.spec.suspend` boolean is quickly changed in Git without making use of the Flux CLI.

> Given this, I am wondering if we could not tackle two issues at once by using `suspend.fluxcd.io/reason` as another way to mark the resource as suspended.

After discussing a bit with @darkowlzz at yesterday's flux bug scrub, I'd like to start an RFC for implementing a suspend-by-annotation and deprecating `.spec.suspend` api.  Perhaps we use `suspend.toolkit.fluxcd.io/state` or `suspend.toolkit.fluxcd.io/suspended` to signal to the controllers that the resource is suspended while allowing an optional reason under `suspend.toolkit.fluxcd.io/reason`.

If that path is supported by the maintainers I would ask that we consider this PR again without any control semantics for the suspend reason annotation.

Reply:
Think `reconcile.fluxcd.io/suspended` would be sufficient to suspend and provide a reason to start with, and fit nicely together with the existing annotation to request a reconciliation (`reconcile.fluxcd.io/requestedAt`).

Reply:
I think adding a suspend annotation is great and RFC is the most reasonable way to kick off a design around it. One thing to keep in mind is that the RFC should be very clear about co-existence with the existing `.spec.suspend` field which we cannot remove in the v1 APIs (e.g. kustomization.v1.kustomize.toolkit.fluxcd.io).

Reply:
I like the suggestion to use `reconcile.fluxcd.io/suspend` annotation for suspending a resource with support for a string that could be anything, including a reason-- maybe "message" is more open.  However if we do want to have this eventually respected by controllers I'm concerned that adding it here will require some later migration.  I'm not in a rush to get this feature PR incorporated and would be happy to get an RFC going to help the community get what they want with minimal churn.

Reply:
After discussing the #4271 at the team meeting in September, 2023 @stefanprodan noted some complications with having suspended state set only as an annotation with a new version cli which would not be recognized by an old version controller.  Possible long term options to support #4271 were suggested by @darkowlzz which included getting object status to indicate suspended state which would probably warrant a separate RFC.

I'd like to request another review of this PR which omits any controller interaction and simply focuses on metadata for human communication.